File: backend/queries.py
"""Выборки сообщений: фильтры ленты, распределение по уровням, окна периодов.
Условия периодов — docs/03-data-model.md, пп. 4 и 5."""
from datetime import datetime, timedelta, timezone

from sqlalchemy import Select, func, or_, select
from sqlalchemy.orm import Session

from backend.models import LEVELS, Connection, Message, User

# Смещение часового пояса приходит с фронта в минутах (как -getTimezoneOffset()),
# иначе «Сегодня» на сервере в UTC не совпадёт с «сегодня» пользователя.
MAX_TZ_OFFSET_MINUTES = 14 * 60  # дальше реальных часовых поясов не бывает


def utcnow() -> datetime:
    return datetime.now(timezone.utc).replace(tzinfo=None)


def period_start(period: str, tz_offset: int = 0, now: datetime | None = None) -> datetime | None:
    """Начало окна фильтра «Период». None — «Всё время»."""
    now = now or utcnow()
    if period == "today":
        minutes = max(-MAX_TZ_OFFSET_MINUTES, min(MAX_TZ_OFFSET_MINUTES, tz_offset))
        offset = timedelta(minutes=minutes)
        local_midnight = (now + offset).replace(hour=0, minute=0, second=0, microsecond=0)
        return local_midnight - offset
    if period == "week":
        return now - timedelta(days=7)
    if period == "month":
        return now - timedelta(days=30)
    return None


def summary_period_start(period: str, now: datetime | None = None) -> datetime:
    """Окно сводки: 24ч / Неделя / Месяц."""
    now = now or utcnow()
    hours = {"24h": 24, "week": 24 * 7, "month": 24 * 30}.get(period, 24)
    return now - timedelta(hours=hours)


def user_messages(user: User) -> Select:
    """Базовая выборка: сообщения всех подключений пользователя."""
    return select(Message).join(Connection).where(Connection.user_id == user.id)


def apply_filters(
    stmt: Select,
    *,
    source: str | None = None,
    level: str | None = None,
    status: str | None = None,
    reply: str | None = None,
    action: str | None = None,
    period: str = "all",
    q: str | None = None,
    tz_offset: int = 0,
    now: datetime | None = None,
) -> Select:
    if source:
        stmt = stmt.where(Connection.kind == source)
    if level and level != "all":
        stmt = stmt.where(Message.effective_level == level)
    if status == "unread":
        stmt = stmt.where(Message.is_read.is_(False))
    elif status == "read":
        stmt = stmt.where(Message.is_read.is_(True))
    elif status == "done":
        stmt = stmt.where(Message.status == "DONE")
    if reply in ("yes", "no"):
        stmt = stmt.where(Message.needs_reply.is_(reply == "yes"))
    if action in ("yes", "no"):
        stmt = stmt.where(Message.needs_action.is_(action == "yes"))
    start = period_start(period, tz_offset, now)
    if start is not None:
        stmt = stmt.where(Message.received_at >= start)
    if q and q.strip():
        # Поиск по отправителю, теме и тексту, регистронезависимо.
        # % и _ внутри запроса — обычные символы, а не шаблон LIKE.
        escaped = q.strip().lower().replace("\\", "\\\\").replace("%", "\\%").replace("_", "\\_")
        pattern = f"%{escaped}%"
        stmt = stmt.where(
            or_(
                func.lower(Message.sender_name).like(pattern, escape="\\"),
                func.lower(Message.sender_addr).like(pattern, escape="\\"),
                func.lower(Message.subject).like(pattern, escape="\\"),
                func.lower(Message.body).like(pattern, escape="\\"),
            )
        )
    return stmt


def level_distribution(db: Session, stmt: Select) -> dict[str, int]:
    """Счётчики по четырём уровням — для полосы распределения. Нули включены."""
    counts = {level: 0 for level in LEVELS}
    subq = stmt.with_only_columns(Message.effective_level.label("lvl")).order_by(None).subquery()
    rows = db.execute(select(subq.c.lvl, func.count()).group_by(subq.c.lvl)).all()
    for level, count in rows:
        if level in counts:
            counts[level] = count
    return counts

File: backend/internal/postgres/queries.go
package postgres

import (
	"strings"
	"time"
)

// MaxTZOffsetMinutes — смещение часового пояса приходит с фронта в минутах
// (как -getTimezoneOffset()). Дальше реальных часовых поясов не бывает.
const MaxTZOffsetMinutes = 14 * 60

// Filter — набор фильтров ленты уровня 2 (docs/03-data-model.md, п. 6).
type Filter struct {
	Source   string
	Level    string
	Status   string
	Reply    string
	Action   string
	Period   string
	Q        string
	TZOffset int
	Now      time.Time
}

// PeriodStart — начало окна фильтра «Период». nil — «Всё время».
func PeriodStart(period string, tzOffset int, now time.Time) *time.Time {
	if now.IsZero() {
		now = UTCNow()
	}
	switch period {
	case "today":
		minutes := tzOffset
		if minutes > MaxTZOffsetMinutes {
			minutes = MaxTZOffsetMinutes
		}
		if minutes < -MaxTZOffsetMinutes {
			minutes = -MaxTZOffsetMinutes
		}
		offset := time.Duration(minutes) * time.Minute
		local := now.Add(offset)
		midnight := time.Date(local.Year(), local.Month(), local.Day(), 0, 0, 0, 0, time.UTC)
		start := midnight.Add(-offset)
		return &start
	case "week":
		start := now.Add(-7 * 24 * time.Hour)
		return &start
	case "month":
		start := now.Add(-30 * 24 * time.Hour)
		return &start
	}
	return nil
}

// SummaryPeriodStart — окно сводки: 24ч / Неделя / Месяц.
func SummaryPeriodStart(period string, now time.Time) time.Time {
	if now.IsZero() {
		now = UTCNow()
	}
	hours := 24
	switch period {
	case "week":
		hours = 24 * 7
	case "month":
		hours = 24 * 30
	}
	return now.Add(-time.Duration(hours) * time.Hour)
}

// escapeLike — % и _ внутри поискового запроса это обычные символы, а не шаблон.
func escapeLike(value string) string {
	value = strings.ReplaceAll(value, `\`, `\\`)
	value = strings.ReplaceAll(value, "%", `\%`)
	return strings.ReplaceAll(value, "_", `\_`)
}

// FilteredMessages — лента уровня 2: сообщения пользователя под фильтрами,
// новые первыми.
func (db *DB) FilteredMessages(userID int64, filter Filter) ([]*Message, error) {
	where := []string{"c.user_id = ?"}
	args := []any{userID}

	if filter.Source != "" {
		where = append(where, "c.kind = ?")
		args = append(args, filter.Source)
	}
	if filter.Level != "" && filter.Level != "all" {
		where = append(where, effectiveLevelSQL+" = ?")
		args = append(args, filter.Level)
	}
	switch filter.Status {
	case "unread":
		where = append(where, "m.is_read = false")
	case "read":
		where = append(where, "m.is_read = true")
	case "done":
		where = append(where, "m.status = 'DONE'")
	}
	if filter.Reply == "yes" || filter.Reply == "no" {
		where = append(where, "m.needs_reply = ?")
		args = append(args, filter.Reply == "yes")
	}
	if filter.Action == "yes" || filter.Action == "no" {
		where = append(where, "m.needs_action = ?")
		args = append(args, filter.Action == "yes")
	}
	if start := PeriodStart(filter.Period, filter.TZOffset, filter.Now); start != nil {
		where = append(where, "m.received_at >= ?")
		args = append(args, start.UTC())
	}
	if query := strings.TrimSpace(filter.Q); query != "" {
		// Поиск по отправителю, теме и тексту, регистронезависимо.
		pattern := "%" + escapeLike(strings.ToLower(query)) + "%"
		where = append(where, `(lower(m.sender_name) LIKE ? ESCAPE '\'
			OR lower(m.sender_addr) LIKE ? ESCAPE '\'
			OR lower(m.subject) LIKE ? ESCAPE '\'
			OR lower(m.body) LIKE ? ESCAPE '\')`)
		args = append(args, pattern, pattern, pattern, pattern)
	}

	return db.queryMessages(
		`SELECT `+messageColumns+` FROM message m JOIN connection c ON c.id = m.connection_id
		 WHERE `+strings.Join(where, " AND ")+`
		 ORDER BY m.received_at DESC, m.id DESC`, args...)
}

// MessagesSince — сообщения пользователя за окно сводки, новые первыми.
func (db *DB) MessagesSince(userID int64, since time.Time) ([]*Message, error) {
	return db.queryMessages(
		`SELECT `+messageColumns+` FROM message m JOIN connection c ON c.id = m.connection_id
		 WHERE c.user_id = ? AND m.received_at >= ?
		 ORDER BY m.received_at DESC, m.id DESC`, userID, since.UTC())
}

// Distribution — счётчики по четырём уровням для полосы распределения.
// Нули включены: полоса всегда состоит из четырёх сегментов.
func Distribution(messages []*Message) map[string]int {
	counts := make(map[string]int, len(Levels))
	for _, level := range Levels {
		counts[level] = 0
	}
	for _, message := range messages {
		counts[message.EffectiveLevel()]++
	}
	return counts
}
